lib/repl.js

```diff
@@ -689,8 +690,31 @@ function intFilter(item) {
   return /^[A-Za-z_$]/.test(item);
 }
 
+const defaultProperties = {
+  ARRAY: Object.getOwnPropertyNames([]).filter(intFilter),
+  BUFFER: Object.getOwnPropertyNames(Buffer.alloc(1)).filter(intFilter)
+};
+
+function mayBeLargeObject(obj) {
+  return (Array.isArray(obj) || Buffer.isBuffer(obj));
+}
+
 function filteredOwnPropertyNames(obj) {
   if (!obj) return [];
+  if (mayBeLargeObject(obj) && obj.length > 1e6) {
+    this._writeToOutput('\r\n');
+    process.emitWarning(
+      'Instance is too large so the completion may missing some custom ' +
+      'properties.',
+      'REPLWarning',
+      undefined,
+      undefined,
+      true);
```

Comment:
@XadillaX fwiw, the autocompletion in REPL is far from perfect and I, as a user, would not and do not expect it to always show all the completions (though it would be an awesome thing). Even IDEs don't do that reliably for JavaScript.